django/db/models/sql/compiler.py

```diff
@@ -454,7 +454,13 @@ def as_sql(self, with_limits=True, with_col_aliases=False):
                 from_, f_params = self.get_from_clause()
                 where, w_params = self.compile(self.where) if self.where is not None else ("", [])
                 having, h_params = self.compile(self.having) if self.having is not None else ("", [])
-                result = ['SELECT']
+                result = []
+                if self.query.explain_query:
+                    prefix = self.connection.ops.explain_query_prefix(
+                        self.query.explain_format,
+                        self.query.explain_verbose)
+                    result.append(prefix)
+                result.append('SELECT')
```

Comment:
Does this mean that support for this is limited to `SELECT` statements?

- PostgreSQL supports `SELECT`, `INSERT`, `UPDATE`, `DELETE`, and others that are irrelevant to Django...
- MySQL support `SELECT`, `INSERT`, `UPDATE`, `DELETE`, `REPLACE`.

Reply:
I wanted to enable this just for `SELECT` queries to begin with. To be honest I'm not sure how I could enable it for the others, `.explain()` would have to be added before an `.update()/delete()` call and would have to modify the returned values. If you have any ideas how this could be done I would love to hear them, but perhaps we can add this as a separate change.

Reply:
Agreed. A follow-up ticket to investigate the possibility of this would be a good idea, along with a discussion on the mailing list.

One way is to implement it as you have said, another could be some sort of context manager approach, e.g.

```python
with queryset.explain() as output:
    updated = queryset.update(field=value)
    print(output)
```

In this case, `updated` ought to be `0`. If `queryset.explain(analyze=True)` were used, `updated` would be the number of records modified (assuming the backend can report this at the same time).

Reply:
Adding hooks in `as_sql` in the different compilers to check whether the query should be executed or explained might be an option? I think it's necessary to do changes in those methods.